### DISCUSSION TITLE
Merge a list of PdfDocument objects into a single PdfDocument object

### DISCUSSION THREAD

**Discussion by hector-sherpas** [pypdfium2-team/pypdfium2#327] [Q&A] [answered]
Is there a way to merge a list of preprocessed PdfDocument objects into a single PdfDocument object? I have tried with import_pages method but it gives an error on the second iteration.

**Comment by mara004** [comment]
Post a reproducible example (code, input file, traceback), then I can take a look.
We can't provide help or debug this with the insufficient information given. FWIW, `import_pages()` works for me.

**Reply by mara004** [reply]
@hector-sherpas Also, guessing from previous interaction that you may be associated with @semoal, please double-check that there aren't any unresolved threading issues (unguarded threaded use of (py)pdfium APIs) on your side before reporting anything. See https://github.com/VikParuchuri/pdftext/pull/7#issuecomment-2331630354
If (py)pdfium's [threading incompatibility](https://pypdfium2.readthedocs.io/en/stable/python_api.html#thread-incompatibility) is not respected, arbitrary issues may arise.

**Reply by hector-sherpas** [reply]
```python
import pypdfium2 as pdfium

def get_doc_page_from_doc(doc: pdfium.PdfDocument) -> list(pdfium.PdfDocument):
    pdf_pages = []
    for page_idx in range(len(doc)):
        blank_doc = pdfium.PdfDocument.new()
        blank_doc.import_pages(doc, pages=[page_idx])
        
    return pdf_pages


def gather_pages_to_unique_doc(pages: list[pdfium.PdfDocument]) -> pdfium.PdfDocument:
    new_doc = pdfium.PdfDocument.new()
    for idx_page, page in enumerate(pages):
        #width, height = page.get_page_size(0)
        #new_doc.new_page(width, height)
        new_doc.import_pages(page, pages=[idx_page])

    return new_doc

doc = pdfium.PdfDocument('path_doc')

pdf_pages = get_doc_page_from_doc(doc)
gather_doc = gather_pages_to_unique_doc(pages)
```
![Screenshot 2024-09-16 at 22 22 11](https://github.com/user-attachments/assets/aa82e6b8-abac-48c9-b83f-3a3c7f60b85e)

**Reply by mara004** [reply, marked as answer]
Looks like `gather_pages_to_unique_doc()`, which triggers the error, might do OOB index access.
You're counting up `idx_page` without checking the document's length.
The `pages` argument of `import_pages()` is, of course, relative to the document being imported, so perhaps you always want the first page (index 0) instead of counting up?

**Reply by mara004** [reply]
Aside from that, calling a variable that holds a PdfDocument `page` is pretty confusing. Maybe `single_page_doc` or something would be clearer. Again assuming that what you want to do is merge a sequence of single-page PdfDocuments back into one PdfDocument.
Note that you can also omit the `pages` argument completely to import all pages of a given doc.

**Reply by hector-sherpas** [reply]
That's right, setting the index to zero or clearing the pages argument merges all the PdfDocument objects in the list into a single PdfDocument  object. Otherwise, it was just a small example. Thanks
